source/mqtt311_decoder.c

```diff
@@ -0,0 +1,206 @@
+
+static void s_aws_mqtt311_decoder_reset(struct aws_mqtt311_decoder *decoder) {
+    decoder->state = AWS_MDST_READ_FIRST_BYTE;
+    decoder->total_packet_length = 0;
+    aws_byte_buf_reset(&decoder->packet_buffer, false);
+}
+
+void aws_mqtt311_decoder_init(
+    struct aws_mqtt311_decoder *decoder,
+    struct aws_allocator *allocator,
+    const struct aws_mqtt311_decoder_options *options) {
+
+    aws_byte_buf_init(&decoder->packet_buffer, allocator, 5);
+    decoder->config = *options;
+
+    s_aws_mqtt311_decoder_reset(decoder);
+}
+
+void aws_mqtt311_decoder_clean_up(struct aws_mqtt311_decoder *decoder) {
+    aws_byte_buf_clean_up(&decoder->packet_buffer);
+}
+
+static void s_aws_mqtt311_decoder_reset_for_new_packet(struct aws_mqtt311_decoder *decoder) {
+    if (decoder->state != AWS_MDST_PROTOCOL_ERROR) {
+        s_aws_mqtt311_decoder_reset(decoder);
+    }
+}
+
+enum aws_mqtt311_decoding_directive { AWS_MDD_CONTINUE, AWS_MDD_OUT_OF_DATA, AWS_MDD_PROTOCOL_ERROR };
+
+static enum aws_mqtt311_decoding_directive aws_result_to_mqtt311_decoding_directive(int result) {
+    return (result == AWS_OP_SUCCESS) ? AWS_MDD_CONTINUE : AWS_MDD_PROTOCOL_ERROR;
+}
+
+static int s_aws_mqtt311_decoder_safe_packet_handle(
+    struct aws_mqtt311_decoder *decoder,
+    enum aws_mqtt_packet_type packet_type,
+    struct aws_byte_cursor packet_cursor) {
+    packet_handler_fn *handler = decoder->config.packet_handlers->handlers_by_packet_type[packet_type];
+    if (handler != NULL) {
+        return handler(packet_cursor, decoder->config.handler_user_data);
+    } else {
+        return aws_raise_error(AWS_ERROR_MQTT_PROTOCOL_ERROR);
+    }
+}
+
+static enum aws_mqtt311_decoding_directive s_handle_decoder_read_first_byte(
+    struct aws_mqtt311_decoder *decoder,
+    struct aws_byte_cursor *data) {
+    AWS_FATAL_ASSERT(decoder->packet_buffer.len == 0);
+    if (data->len == 0) {
+        return AWS_MDD_OUT_OF_DATA;
+    }
+
+    /*
+     * Do a greedy check to see if the whole MQTT packet is contained within the received data.  If it is, decode it
+     * directly from the incoming data cursor without buffering it first.  Otherwise, the packet is fragmented
+     * across multiple received data calls, and so we must use the packet buffer and copy everything first via the
+     * full decoder state machine.
+     *
+     * A corollary of this is that the decoder is only ever in the AWS_MDST_READ_REMAINING_LENGTH or AWS_MDST_READ_BODY
+     * states if the current MQTT packet was received in a fragmented manner.
+     */
+    struct aws_byte_cursor temp_cursor = *data;
+    struct aws_mqtt_fixed_header packet_header;
+    AWS_ZERO_STRUCT(packet_header);
+    if (!aws_mqtt_fixed_header_decode(&temp_cursor, &packet_header) &&
+        temp_cursor.len >= packet_header.remaining_length) {
+
+        /* figure out the cursor that spans the full packet */
+        size_t fixed_header_length = temp_cursor.ptr - data->ptr;
```

Comment:
I'm don't think that's equivalent, data is unchanged at this point

Reply:
whatever, if my suggestion is confusing too, then it's not much better

Reply:
The `data` is unchanged, while `temp_cursor.len` has been updated since we proceed the header here. It seems that the difference of `data.len` and `temp_cursor.len` should be the same as the header length here.

Reply:
NM I see it now.